File: libs/power-log-parser/src/lib/parsers/mercenaries-selected-target-parser.ts
import { GameTag } from '@firestone-hs/reference-data';
import { ActionParser } from '../action-parser';
import { GameEventProvider, GameEventHelper } from '../game-event';
import { Node, NodeType, TagChange } from '../models';
import { GameState } from '../state/game-state';
import { ParserState, StateType } from '../state/parser-state';
import type { StateFacade } from '../state/state-facade';

export class MercenariesSelectedTargetParser implements ActionParser {
	readonly ParserName = 'MercenariesSelectedTargetParser';

	private GameState: GameState;
	private StateFacade: StateFacade;

	constructor(parserState: ParserState, stateFacade: StateFacade) {
		this.GameState = parserState.GameState;
		this.StateFacade = stateFacade;
	}

	AppliesOnNewNode(node: Node, stateType: StateType): boolean {
		return (
			stateType === StateType.PowerTaskList &&
			node.Type === NodeType.TagChange &&
			(node.Object as TagChange).Name === (GameTag.LETTUCE_SELECTED_TARGET as number)
		);
	}

	AppliesOnCloseNode(_node: Node, _stateType: StateType): boolean {
		return false;
	}

	CreateGameEventProviderFromNew(node: Node): GameEventProvider[] | null {
		const tagChange = node.Object as TagChange;
		const entity = this.GameState.CurrentEntities.get(tagChange.Entity)!;
		const cardId = entity.CardId;
		const controllerId = entity.GetEffectiveController();

		if (tagChange.Value === 0) {
			return [
				GameEventProvider.Create(
					tagChange.TimeStamp,
					'MERCENARIES_UNSELECTED_TARGET',
					GameEventHelper.CreateProvider(
						'MERCENARIES_UNSELECTED_TARGET',
						cardId,
						controllerId,
						entity.Id,
						this.StateFacade,
						null,
					),
					true,
					node,
				),
			];
		} else {
			const targetEntityId = tagChange.Value;
			const targetEntity = this.GameState.CurrentEntities.get(targetEntityId)!;
			const targetCardId = this.GameState.GetCardIdForEntity(targetEntityId);
			const targetControllerId = targetEntity.GetEffectiveController();

			return [
				GameEventProvider.Create(
					tagChange.TimeStamp,
					'MERCENARIES_SELECTED_TARGET',
					GameEventHelper.CreateProvider(
						'MERCENARIES_SELECTED_TARGET',
						cardId,
						controllerId,
						entity.Id,
						this.StateFacade,
						{
							TargetControllerId: targetControllerId,
							TargetCardId: targetCardId,
							TargetEntityId: targetEntityId,
						},
					),
					true,
					node,
				),
			];
		}
	}

	CreateGameEventProviderFromClose(_node: Node): GameEventProvider[] | null {
		return null;
	}
}

File: tools/linux-helpers/hs-converter/HearthstoneReplays/Events/Parsers/MercenariesSelectedTargetParser.cs
using HearthstoneReplays.Parser;
using HearthstoneReplays.Parser.ReplayData;
using HearthstoneReplays.Parser.ReplayData.GameActions;
using System;
using HearthstoneReplays.Enums;
using HearthstoneReplays.Parser.ReplayData.Entities;
using System.Collections.Generic;

namespace HearthstoneReplays.Events.Parsers
{
    public class MercenariesSelectedTargetParser : ActionParser
    {
        private GameState GameState { get; set; }
        private ParserState ParserState { get; set; }
        private StateFacade StateFacade { get; set; }

        public MercenariesSelectedTargetParser(ParserState ParserState, StateFacade stateFacade)
        {
            this.ParserState = ParserState;
            this.GameState = ParserState.GameState;
            this.StateFacade = stateFacade;
        }

        public bool AppliesOnNewNode(Node node, StateType stateType)
        {
            TagChange tagChange = null;
            return stateType == StateType.PowerTaskList
                && node.Type == typeof(TagChange)
                && ((tagChange = node.Object as TagChange).Name == (int)GameTag.LETTUCE_SELECTED_TARGET);
        }

        public bool AppliesOnCloseNode(Node node, StateType stateType)
        {
            return false;
        }

        public List<GameEventProvider> CreateGameEventProviderFromNew(Node node)
        {
            var tagChange = node.Object as TagChange;
            var entity = GameState.CurrentEntities[tagChange.Entity];
            var cardId = entity.CardId;
            var controllerId = entity.GetEffectiveController();

            if (tagChange.Value == 0){
                return new List<GameEventProvider> { GameEventProvider.Create(
                    tagChange.TimeStamp,
                    "MERCENARIES_UNSELECTED_TARGET",
                    GameEvent.CreateProvider(
                        "MERCENARIES_UNSELECTED_TARGET",
                        cardId,
                        controllerId,
                        entity.Id,
                        StateFacade,
                        null),
                    true,
                    node) };
            }else {
                var targetEntityId = tagChange.Value;
                var targetEntity = GameState.CurrentEntities[targetEntityId];
                var targetCardId = GameState.GetCardIdForEntity(targetEntityId);
                var targetControllerId = targetEntity.GetEffectiveController();

                return new List<GameEventProvider> { GameEventProvider.Create(
                    tagChange.TimeStamp,
                    "MERCENARIES_SELECTED_TARGET",
                    GameEvent.CreateProvider(
                        "MERCENARIES_SELECTED_TARGET",
                        cardId,
                        controllerId,
                        entity.Id,
                        StateFacade,
                        //null,
                        new {
                            TargetControllerId = targetControllerId,
                            TargetCardId = targetCardId,
                            TargetEntityId = targetEntityId
                        }),
                    true,
                    node) };
            }
            
       
        }

        public List<GameEventProvider> CreateGameEventProviderFromClose(Node node)
        {
            return null;
        }
    }
}
